### ISSUE TITLE
Force a Host HTTP header to remote

### ISSUE THREAD

**Issue by arnaudmz** [esigate/esigate#88]
Instead of simply having a preserveHost=(true|false) to forward the recived Host header variable or reuse the remoteBaseUrl,
it would be great to be able to **force an arbitrary Host** header value to be sent to the remote app

* Today, option1
```
# public VHost: plop.org
plop.remoteBaseUrl=http://10.0.0.1
plop.preserveHost=true
```
=> will send **Host: plop.org** to remote 10.0.0.1

* Today, **option2**

```
# public VHost: plop.org
plop.remoteBaseUrl=http://10.0.0.1
plop.preserveHost=false
```
=> will send **Host: 10.0.0.1** to remote 10.0.0.1

* What would be nice: 
```
# public VHost: plop.org
plop.remoteBaseUrl=http://10.0.0.1
plop.preserveHost=false
plop.forceHost=my-fake-vhost.org
```
=> will send **Host: my-fake-vhost.org** to remote 10.0.0.1

**Comment by nricheton** [comment]
Hi Arnaud,

Did you try :

# public VHost: plop.org
plop.remoteBaseUrl=http://my-fake-vhost.org
plop.preserveHost=false

I think this would work out of the box.

Another option would be to use an extension and replace the Host header before sending the request (event: pre fragment) (should be ok in a few lines of code)

Anyway, in my opinion is a -1 for adding this to esigate . Extensions should be used for non-standard features.

**Comment by arnaudmz** [comment]
Actually if you suggest to add some kind of alias into /etc/hosts  (or in DNS) such as
'''
10.0.0.1 my-fake-vhost.org
'''
Of course it works and BTW it sounds to be the only way to do the job for now.

But this looks to me like a very bad idea to fake the local host resolution to lead a public FQDN to a private IP.
In this case the internal resolution of my-fake-vhost.org on the ESIGate host would be different from the external resolution for some technical reasons.

Rewriting the Host header when forwarding trafic to remotes is pretty standard in nginx and haproxy and would avoid this.

**Comment by nricheton** [comment]
I thought the dns was available in your case.

Anyway, the good way to do it is to use an extension to rewrite the host header.
It's only a single line : 

```
((FragmentEvent) event).getHttpRequest().setHeader("Host", "my-fake-vhost.org");
```

Please take a look at :
https://github.com/esigate/esigate/blob/master/esigate-core/src/main/java/org/esigate/extension/ForwardOriginalUrl.java

In esigate 2.x and 3.x we had a lot of requests for project-specific behaviors and we were adding a lot of configuration options.
We now have a core with standard features (HTTP, link rewriting) and extensions for all custom / improved features.

**Comment by fxbonnet** [comment]
Hi Nicolas,

I don't think it is as simple as setting the Host header because there is already a hack in the core about how Apache HttpClient handles URI, route resolution, host header and also the cache key. There is also an impact on URL rewriting as the target server will likely use the forced host name in the URLs in response headers or inside html pages.

I already had the need for such a feature on several projects, we always solved this by changing the /etc/hosts of the machines but sys administrators were not very happy with that.

For all there reasons I think this feature should be a core feature.

**Comment by alexist** [comment]
In Apache httpclient, connection manager can be configured with a DNS Resolver. By default, the DNS resolver uses the default OS implementation for resolving host names. But we can create a Custom DNS Resolver that first lookup in a collection held in-memory per provider, then uses the default OS implementation.

Esigate configuration would be :
# public VHost: plop.org
plop.remoteBaseUrl=http://my-fake-vhost.org
plop.preserveHost=false
plop.remoteIP=10.0.0.1

**Comment by fxbonnet** [comment]
+1

**Comment by nricheton** [comment]
+1 
But please implement this by providing an extension point for HttpClient creation 
(issue #133 ) and create an extension for this (issue #88 ). 

I understand some projects may need this, but this is the result of a special requirement (sending a host header which is not the host you use and/or the the one the user have requested). 

In the past, most projects had a special use case which required changes to esigate core, and we applied those changes. As the result the code got more complex and more lines/branches were run even if you didn't need them on all other projects.

We have now a modular architecture which provide a clean separation on the core and specific needs. I think we better extend its scope (with issue #133) and provide this feature in esigate-core, but by using the extension mechanism .

**Comment by alexist** [comment]
Hi,

The DNS extension is implemented https://github.com/alexist/esigate/tree/HttpBuilderCustomization_133.

As HTTPClientBuilder didn't expose DNSResolver, i made a patch that has been push to 4.5.2-SNAPSHOT. The version 4.5.2 will be release in february.

If you want to test, you have to install httpclient 4.5.2-SNAPSHOT.

**Comment by nricheton** [comment]
Looks good.

**Comment by alexist** [comment]
Fixed in 5.2
